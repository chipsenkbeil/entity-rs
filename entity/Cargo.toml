[package]
name = "entity"
description = "Library that provides entity-like constructs"
categories = ["data-structures"]
version = "0.1.0"
authors = ["Chip Senkbeil <chip@senkbeil.org>"]
edition = "2018"
homepage = "https://github.com/chipsenkbeil/entity-rs"
repository = "https://github.com/chipsenkbeil/entity-rs"
readme = "README.md"
license = "MIT OR Apache-2.0"

[features]
full = ["async-graphql", "global", "macros", "inmemory_db", "serde-1", "sled_db"]
global = ["lazy_static"]
inmemory_db = []
macros = ["entity_macros"]
serde-1 = ["serde", "serde/rc", "typetag"]
sled_db = ["bincode", "serde-1", "sled"]

[dependencies]
derivative = "2.1.1"
derive_more = { version = "0.99.11", default-features = false, features = ["as_ref", "as_mut", "constructor", "deref", "deref_mut", "display", "error", "from", "into", "into_iterator", "try_into"] }
doc-comment = "0.3.3"
dyn-clone = "1.0.3"
paste = "1.0.4"
strum = { version = "0.19", features = ["derive"] }
entity_noop_macros = { version = "0.1.0", path = "../entity_noop_macros" }

async-graphql = { version = "2.4.6", optional = true }
bincode = { version = "1.3.1", optional = true }
lazy_static = { version = "1.4.0", optional = true }
serde = { version = "1.0.117", features = ["derive"], optional = true }
sled = { version = "0.34.6", optional = true }
typetag = { version = "0.1.6", optional = true }
entity_macros = { version = "0.1.0", path = "../entity_macros", optional = true }

[dev-dependencies]
futures = "0.3.9"

[build-dependencies]
# Using -Z minimal-versions fails with the async-graphql dependency as it
# has the follow dep chain of
#
# multer 1.2.2 -> mime 0.3.0 -> unicase 2.0.0 -> rustc_version 0.1.0 -> semver 0.1.0
#
# The semver version fails to compile and is ancient, so we are providing a
# requirement on a later version of the build dependency to enable using
# the minimal version check
semver = "0.11.0"
