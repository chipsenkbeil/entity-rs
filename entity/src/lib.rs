#![forbid(unsafe_code)]

mod alloc;
mod any;
mod database;
mod ent;
pub mod ext;
pub mod global;

pub use alloc::{Id, IdAllocator, EPHEMERAL_ID};
pub use any::*;
pub use database::*;
pub use ent::*;

#[cfg(feature = "macros")]
pub use entity_macros::*;

/// Represents a collection of core traits that should be imported to enable
/// methods on a variety of types such as [`Ent`] and [`EntBuilder`]
pub mod prelude {
    pub use super::{
        AsAny, Database, DatabaseExt, Ent, EntBuilder, EntExt, EntLoader, EntQuery, EntType,
        EntWrapper,
    };
}

/// Vendor module to re-expose relevant libraries
pub mod vendor {
    #[cfg(feature = "sled_db")]
    pub use ::sled;

    /// Re-exported macros, useful only to [`entity_macros`] crate
    pub mod macros {
        /// Re-export of serde
        pub mod serde {
            /// Indicates whether or not the included serde derive macros are
            /// the result of the feature existing (true) or a no-op (false)
            #[inline]
            pub const fn exists() -> bool {
                cfg!(feature = "serde-1")
            }

            #[cfg(feature = "serde-1")]
            pub use ::serde::Serialize;

            #[cfg(feature = "serde-1")]
            pub use ::serde::Deserialize;

            #[cfg(not(feature = "serde-1"))]
            pub use ::entity_noop_macros::NoopDeriveSerde as Serialize;

            #[cfg(not(feature = "serde-1"))]
            pub use ::entity_noop_macros::NoopDeriveSerde as Deserialize;
        }

        /// Re-export of typetag
        pub mod typetag {
            /// Indicates whether or not the included typetag attr macro is the
            /// result of the feature existing (true) or a no-op (false)
            #[inline]
            pub const fn exists() -> bool {
                cfg!(feature = "typetag")
            }

            #[cfg(feature = "typetag")]
            pub use ::typetag::serde;

            #[cfg(not(feature = "typetag"))]
            pub use ::entity_noop_macros::noop_attr as serde;
        }
    }
}
