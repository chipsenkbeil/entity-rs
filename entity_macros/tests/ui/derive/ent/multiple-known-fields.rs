//! Test that trying to declare multiple ent-fields on a single field fails properly.
//!
//! This test **must not** claim that `TestEnt` is missing any required `ent` fields;
//! if it does, that likely indicates a bug where field analysis is bailing out too early.

use entity::{Ent, Id, WeakDatabaseRc};

#[derive(Clone, Ent)]
struct TestEnt {
    #[ent(id, database)]
    id: Id,

    #[ent(database, id)]
    database: WeakDatabaseRc,

    #[ent(created, last_updated)]
    created: u64,

    #[ent(last_updated, id)]
    last_updated: u64,
}

fn main() {}