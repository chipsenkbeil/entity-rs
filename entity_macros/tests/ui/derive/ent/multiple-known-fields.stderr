error: A field can only declare one of `id`, `database`, `created`, or `last_updated`
  --> $DIR/multiple-known-fields.rs:10:11
   |
10 |     #[ent(id, database)]
   |           ^^

error: A field can only declare one of `id`, `database`, `created`, or `last_updated`
  --> $DIR/multiple-known-fields.rs:10:15
   |
10 |     #[ent(id, database)]
   |               ^^^^^^^^

error: A field can only declare one of `id`, `database`, `created`, or `last_updated`
  --> $DIR/multiple-known-fields.rs:13:21
   |
13 |     #[ent(database, id)]
   |                     ^^

error: A field can only declare one of `id`, `database`, `created`, or `last_updated`
  --> $DIR/multiple-known-fields.rs:13:11
   |
13 |     #[ent(database, id)]
   |           ^^^^^^^^

error: Already have an id elsewhere
  --> $DIR/multiple-known-fields.rs:14:5
   |
14 |     database: WeakDatabaseRc,
   |     ^^^^^^^^

error: Already have a database elsewhere
  --> $DIR/multiple-known-fields.rs:14:5
   |
14 |     database: WeakDatabaseRc,
   |     ^^^^^^^^

error: A field can only declare one of `id`, `database`, `created`, or `last_updated`
  --> $DIR/multiple-known-fields.rs:16:11
   |
16 |     #[ent(created, last_updated)]
   |           ^^^^^^^

error: A field can only declare one of `id`, `database`, `created`, or `last_updated`
  --> $DIR/multiple-known-fields.rs:16:20
   |
16 |     #[ent(created, last_updated)]
   |                    ^^^^^^^^^^^^

error: A field can only declare one of `id`, `database`, `created`, or `last_updated`
  --> $DIR/multiple-known-fields.rs:19:25
   |
19 |     #[ent(last_updated, id)]
   |                         ^^

error: A field can only declare one of `id`, `database`, `created`, or `last_updated`
  --> $DIR/multiple-known-fields.rs:19:11
   |
19 |     #[ent(last_updated, id)]
   |           ^^^^^^^^^^^^

error: Already have an id elsewhere
  --> $DIR/multiple-known-fields.rs:20:5
   |
20 |     last_updated: u64,
   |     ^^^^^^^^^^^^

error: Already have a last_updated timestamp elsewhere
  --> $DIR/multiple-known-fields.rs:20:5
   |
20 |     last_updated: u64,
   |     ^^^^^^^^^^^^
